pkg/pquerier/frames.go

```diff
@@ -0,0 +1,660 @@
+}
+
+func (c *basicColumn) SetDataAt(i int, value interface{}) error { return nil }
+func (c *basicColumn) SetData(d interface{}, size int) error {
+	return errors.New("method not supported")
+}
+
+// DType is data type
+type DType reflect.Type
+
+func NewDataColumn(name string, colSpec columnMeta, size int, datatype DType) *dataColumn {
+	dc := &dataColumn{basicColumn: basicColumn{name: name, spec: colSpec, size: size}}
+	dc.initializeData(datatype)
+	dc.interpolationFunction = GetInterpolateFunc(colSpec.interpolationType)
+	dc.interpolationTolerance = colSpec.interpolationTolerance
+	return dc
+
+}
+
+type dataColumn struct {
+	basicColumn
+	data                   interface{}
+	interpolationFunction  InterpolationFunction
+	interpolationTolerance int64
+}
+
+func (dc *dataColumn) GetInterpolationFunction() (InterpolationFunction, int64) {
+	return dc.interpolationFunction, dc.interpolationTolerance
+}
+
+func (dc *dataColumn) initializeData(dataType DType) {
+	switch dataType {
+	case IntType:
+		dc.data = make([]int64, dc.size)
+	case FloatType:
+		dc.data = make([]float64, dc.size)
+	case StringType:
+		dc.data = make([]string, dc.size)
+	case TimeType:
+		dc.data = make([]time.Time, dc.size)
+	case BoolType:
+		dc.data = make([]bool, dc.size)
+	}
+}
+
+// DType returns the data type
+func (dc *dataColumn) DType() DType {
+	return reflect.TypeOf(dc.data)
+}
+
+// FloatAt returns float64 value at index i
+func (dc *dataColumn) FloatAt(i int) (float64, error) {
+	if !dc.isValidIndex(i) {
+		return 0, fmt.Errorf("index %d out of bounds [0:%d]", i, dc.size)
+	}
+
+	typedCol, ok := dc.data.([]float64)
+	if !ok {
+		return 0, fmt.Errorf("wrong type (type is %s)", dc.DType())
+	}
+
+	return typedCol[i], nil
+}
+
+// StringAt returns string value at index i
+func (dc *dataColumn) StringAt(i int) (string, error) {
+	if !dc.isValidIndex(i) {
+		return "", fmt.Errorf("index %d out of bounds [0:%d]", i, dc.size)
+	}
+
+	typedCol, ok := dc.data.([]string)
+	if !ok {
+		return "", fmt.Errorf("wrong type (type is %s)", dc.DType())
+	}
+
+	return typedCol[i], nil
+}
+
+// TimeAt returns time.Time value at index i
+func (dc *dataColumn) TimeAt(i int) (int64, error) {
+	if !dc.isValidIndex(i) {
+		return 0, fmt.Errorf("index %d out of bounds [0:%d]", i, dc.size)
+	}
+
+	typedCol, ok := dc.data.([]int64)
+	if !ok {
+		return 0, fmt.Errorf("wrong type (type is %s)", dc.DType())
+	}
+
+	return typedCol[i], nil
+}
+
+func (dc *dataColumn) SetData(d interface{}, size int) error {
+	dc.data = d
+	dc.size = size
+	return nil
+}
+
+func (dc *dataColumn) SetDataAt(i int, value interface{}) error {
+	if !dc.isValidIndex(i) {
+		return fmt.Errorf("index %d out of bounds [0:%d]", i, dc.size)
+	}
+
+	switch reflect.TypeOf(dc.data) {
+	case IntType:
+		dc.data.([]int64)[i] = value.(int64)
+	case FloatType:
+		// Update requested cell, only if not trying to override an existing value with NaN
+		if !(math.IsNaN(value.(float64)) && dc.data.([]float64)[i] != 0) {
```

Comment:
Pay attention that `0` is a valid value. Maybe you need to initialize the data arrays with `NaN`